src/cargo/sources/git/utils.rs

```diff
@@ -855,51 +856,123 @@ pub fn fetch(
-        // recover from corrupt repositories, but we don't want Cargo to stomp
-        // over other legitimate errors.
-        //
-        // Consequently we save off the error of the `fetch` operation and if it
-        // looks like a "corrupt repo" error then we blow away the repo and try
-        // again. If it looks like any other kind of error, or if we've already
-        // blown away the repository, then we want to return the error as-is.
-        let mut repo_reinitialized = false;
-        loop {
-            debug!("initiating fetch of {:?} from {}", refspecs, url);
-            let res = repo
-                .remote_anonymous(url)?
-                .fetch(&refspecs, Some(&mut opts), None);
-            let err = match res {
-                Ok(()) => break,
-                Err(e) => e,
-            };
-            debug!("fetch failed: {}", err);
-
-            if !repo_reinitialized && matches!(err.class(), ErrorClass::Reference | ErrorClass::Odb)
-            {
-                repo_reinitialized = true;
-                debug!(
-                    "looks like this is a corrupt repository, reinitializing \
+    if config
+        .cli_unstable()
+        .gitoxide
+        .map_or(false, |git| git.fetch)
+    {
+        use git::remote::fetch::Error;
+        use git_repository as git;
+        let git2_repo = repo;
+        oxide::with_retry_and_progress(
+            &git2_repo.path().to_owned(),
+            config,
+            &mut |repo, should_interrupt, progress| {
+                // The `fetch` operation here may fail spuriously due to a corrupt
+                // repository. It could also fail, however, for a whole slew of other
+                // reasons (aka network related reasons). We want Cargo to automatically
+                // recover from corrupt repositories, but we don't want Cargo to stomp
+                // over other legitimate errors.
+                //
+                // Consequently we save off the error of the `fetch` operation and if it
+                // looks like a "corrupt repo" error then we blow away the repo and try
+                // again. If it looks like any other kind of error, or if we've already
+                // blown away the repository, then we want to return the error as-is.
+                let mut repo_reinitialized = false;
+                let mut repo_storage;
+                let mut repo = &*repo;
+                loop {
+                    debug!("initiating fetch of {:?} from {}", refspecs, url);
+                    let res = repo
+                        .remote_at(url)?
+                        .with_refspecs(
+                            refspecs.iter().map(|s| s.as_str()),
+                            git::remote::Direction::Fetch,
+                        )?
+                        .connect(git::remote::Direction::Fetch, progress.add_child("fetch"))?
+                        .prepare_fetch(git::remote::ref_map::Options::default())?
```

Comment:
I agree. And I also think that the usecase covered by `git2` is one that could have better support in `gitoxide`  as well. Right now the only choice one has is to use `git::open::Option::isolated()` as it will only read repository configuration which usually doesn't contain more general settings that you might not want. Right now, one has to know (probably by reading documentation) which configuration is affecting a particular call, and override it specifically. All this is quite error prone.

Instead, `gitoxide` could provide a filter for configuration sections and/or keys that should be allowed. That way, it could read the entire configuration which is usually needed to read important values, like `credential.helper`, but would omit everything else which then defaults to standard `git` values (most of the time, `gitoxide` has different defaults if there is the choice between single- and multi-threading, choosing the latter by default).

A last stage of this 'adding explicit control' could be to add builders or options that would override all other configuration it would pull from git configuration.

Yes, let's wait and see to learn if it's more like a feature, or indeed a shortcoming.

Reply:
Something else I hadn't considered before is what are the performance implications of this?  If each call has to look up a load of configuration, that seems wasteful.  Similar for having to deal with a stringly typed API for overriding.  Even if there is caching, that is hidden from the user and they might do things suboptimally (for example passing around paths and constructing new `Repository`s each time).

Granted, this is theoretical, in a lot of cases the overhead might not be noticeable with all of the filesystem operations.

Reply:
I agree, opening a repo once per call is more wasteful than being able to cache and reuse it. The way this is configured, it will consider a couple of files and load them.

Within the `cargo` codebase, this is temporary as one day `git::Repository` can be where `git2::Repository` is now which should avoid some unnecessary `open()` calls.

And even if one would open a `git::Repository`, I was pleasantly surprise to hear [that opening Repos is **significantly** faster](https://github.com/stacked-git/stgit/discussions/255) than it is with `git2`. It's crazy, but not entirely unexpected as `git::Repository` is as lazy as it can be, doing only the absolutely necessary work to correctly open a repository. That shows, and with what cargo does and the couple of Repos it might open, I don't think this will be visible at all (and if it is, it's temporary and goes away once `git2` has been replaced).